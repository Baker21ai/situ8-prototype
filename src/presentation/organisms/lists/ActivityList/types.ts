/**
 * TypeScript interfaces for ActivityList compound component
 */

import { EnterpriseActivity, ActivityCluster } from '../../../../../lib/types/activity';
import { Priority } from '../../../../../lib/utils/status';
import { EnterpriseCardVariant } from '../../../../../components/EnterpriseActivityCard';

// Filter state for activities
export interface ActivityFilterState {
  search: string;
  priorities: Priority[];
  statuses: string[];
  types: string[];
  locations: string[];
  timeRange: 'live' | '15m' | '1h' | '4h' | '24h' | 'custom';
  businessImpact: string[];
  confidenceThreshold: number;
  showClusters: boolean;
  aiFiltering: boolean;
}

// Performance metrics
export interface ActivityPerformanceMetrics {
  totalActivities: number;
  renderTime: number;
  filteredCount: number;
  criticalCount: number;
  averageConfidence: number;
  processingRate: number;
}

// Layout modes
export type ActivityLayoutMode = 'grid' | 'horizontal';
export type ActivityViewMode = 'stream' | 'minimal' | 'summary';

// Context data for compound component
export interface ActivityListContextData {
  // Data
  activities: (EnterpriseActivity | ActivityCluster)[];
  filteredActivities: (EnterpriseActivity | ActivityCluster)[];
  
  // State
  filters: ActivityFilterState;
  selectedItems: Set<string>;
  viewMode: ActivityViewMode;
  layoutMode: ActivityLayoutMode;
  compactMode: boolean;
  useVirtualScrolling: boolean;
  showPerformanceMetrics: boolean;
  
  // Metrics
  performanceMetrics: ActivityPerformanceMetrics;
  priorityMetrics: Record<Priority, number>;
  
  // Event handlers
  onActivitySelect?: (activity: EnterpriseActivity | ActivityCluster) => void;
  onActivityAction?: (action: string, activity: EnterpriseActivity | ActivityCluster) => void;
  onBulkAction?: (action: string, activities: EnterpriseActivity[]) => void;
  
  // State setters
  setFilters: (filters: Partial<ActivityFilterState>) => void;
  setSelectedItems: (items: Set<string>) => void;
  setViewMode: (mode: ActivityViewMode) => void;
  setLayoutMode: (mode: ActivityLayoutMode) => void;
  setCompactMode: (compact: boolean) => void;
  setUseVirtualScrolling: (enabled: boolean) => void;
  setShowPerformanceMetrics: (show: boolean) => void;
  
  // Derived data
  hasSelection: boolean;
  isEmpty: boolean;
  
  // Configuration
  realTimeMode?: boolean;
  height?: number;
  className?: string;
}

// Component props
export interface ActivityListProps {
  activities: (EnterpriseActivity | ActivityCluster)[];
  onActivitySelect?: (activity: EnterpriseActivity | ActivityCluster) => void;
  onActivityAction?: (action: string, activity: EnterpriseActivity | ActivityCluster) => void;
  onBulkAction?: (action: string, activities: EnterpriseActivity[]) => void;
  realTimeMode?: boolean;
  className?: string;
  height?: number;
  children?: React.ReactNode;
}

// Header component props
export interface ActivityListHeaderProps {
  children?: React.ReactNode;
  className?: string;
}

// Content component props
export interface ActivityListContentProps {
  className?: string;
  height?: number;
}

// Footer component props
export interface ActivityListFooterProps {
  className?: string;
}

// Filters component props
export interface ActivityListFiltersProps {
  className?: string;
  showAdvanced?: boolean;
}

// Stats component props
export interface ActivityListStatsProps {
  className?: string;
  detailed?: boolean;
}

// Search component props
export interface ActivityListSearchProps {
  placeholder?: string;
  className?: string;
}

// View toggle component props
export interface ActivityListViewToggleProps {
  className?: string;
}

// Priority segment for display
export interface PrioritySegment {
  priority: Priority;
  items: (EnterpriseActivity | ActivityCluster)[];
  count: number;
  color: {
    bg: string;
    border: string;
    text: string;
  };
  icon: React.ReactNode;
}

// Compound component type with sub-components
export interface ActivityListCompound {
  (props: ActivityListProps): JSX.Element;
  Header: React.ComponentType<ActivityListHeaderProps>;
  Content: React.ComponentType<ActivityListContentProps>;
  Footer: React.ComponentType<ActivityListFooterProps>;
  Filters: React.ComponentType<ActivityListFiltersProps>;
  Stats: React.ComponentType<ActivityListStatsProps>;
  Search: React.ComponentType<ActivityListSearchProps>;
  ViewToggle: React.ComponentType<ActivityListViewToggleProps>;
}