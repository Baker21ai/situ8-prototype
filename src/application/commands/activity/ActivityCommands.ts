/**
 * Activity Domain Commands
 * All commands related to activity management
 */

import { ICommand } from '../base/ICommand';
import { ActivityType } from '../../../lib/utils/security';
import { Priority, Status } from '../../../lib/utils/status';

// ===== CREATE ACTIVITY COMMAND =====

export interface CreateActivityCommand extends ICommand {
  readonly type: 'CreateActivity';
  readonly data: {
    activityType: ActivityType;
    title: string;
    location: string;
    priority: Priority;
    description?: string;
    building?: string;
    zone?: string;
    assignedTo?: string;
    confidence?: number;
    externalData?: {
      sourceSystem: string;
      originalType: string;
      rawPayload: Record<string, any>;
    };
  };
}

export interface CreateActivityResult {
  activityId: string;
  autoTagsApplied: string[];
  incidentTriggered?: {
    incidentId: string;
    rule: string;
  };
}

// ===== UPDATE ACTIVITY COMMAND =====

export interface UpdateActivityCommand extends ICommand {
  readonly type: 'UpdateActivity';
  readonly aggregateId: string; // Activity ID
  readonly data: {
    updates: {
      priority?: Priority;
      status?: Status;
      description?: string;
      assignedTo?: string;
      userTags?: string[];
    };
    reason?: string;
  };
}

export interface UpdateActivityResult {
  updatedFields: string[];
  triggeredWorkflows?: string[];
}

// ===== ASSIGN ACTIVITY COMMAND =====

export interface AssignActivityCommand extends ICommand {
  readonly type: 'AssignActivity';
  readonly aggregateId: string; // Activity ID
  readonly data: {
    assignedTo: string;
    assignedBy: string;
    reason?: string;
    notifyAssignee?: boolean;
  };
}

export interface AssignActivityResult {
  previousAssignee?: string;
  notificationSent: boolean;
}

// ===== ARCHIVE ACTIVITY COMMAND =====

export interface ArchiveActivityCommand extends ICommand {
  readonly type: 'ArchiveActivity';
  readonly aggregateId: string; // Activity ID
  readonly data: {
    reason: string;
    archivedBy: string;
    permanent?: boolean;
  };
}

export interface ArchiveActivityResult {
  archivedAt: Date;
  retentionDate: Date;
}

// ===== BULK STATUS UPDATE COMMAND =====

export interface BulkUpdateStatusCommand extends ICommand {
  readonly type: 'BulkUpdateStatus';
  readonly data: {
    activityIds: string[];
    status: Status;
    reason?: string;
    notifyAssigned?: boolean;
  };
}

export interface BulkUpdateStatusResult {
  successCount: number;
  failedIds: string[];
  notificationsSent: number;
}

// ===== ADD USER TAG COMMAND =====

export interface AddUserTagCommand extends ICommand {
  readonly type: 'AddUserTag';
  readonly aggregateId: string; // Activity ID
  readonly data: {
    tag: string;
    addedBy: string;
  };
}

export interface AddUserTagResult {
  tag: string;
  allTags: string[];
}

// ===== REMOVE USER TAG COMMAND =====

export interface RemoveUserTagCommand extends ICommand {
  readonly type: 'RemoveUserTag';
  readonly aggregateId: string; // Activity ID
  readonly data: {
    tag: string;
    removedBy: string;
  };
}

export interface RemoveUserTagResult {
  tag: string;
  allTags: string[];
}

// ===== ESCALATE ACTIVITY COMMAND =====

export interface EscalateActivityCommand extends ICommand {
  readonly type: 'EscalateActivity';
  readonly aggregateId: string; // Activity ID
  readonly data: {
    escalationType: 'priority' | 'assignment' | 'incident';
    targetPriority?: Priority;
    targetAssignee?: string;
    createIncident?: boolean;
    reason: string;
    escalatedBy: string;
  };
}

export interface EscalateActivityResult {
  escalationType: string;
  previousState: {
    priority?: Priority;
    assignedTo?: string;
  };
  newState: {
    priority?: Priority;
    assignedTo?: string;
    incidentId?: string;
  };
}

// ===== LINK TO INCIDENT COMMAND =====

export interface LinkToIncidentCommand extends ICommand {
  readonly type: 'LinkToIncident';
  readonly aggregateId: string; // Activity ID
  readonly data: {
    incidentId: string;
    linkType: 'trigger' | 'related' | 'evidence';
    linkedBy: string;
    notes?: string;
  };
}

export interface LinkToIncidentResult {
  incidentId: string;
  linkType: string;
  allIncidentLinks: string[];
}

// ===== BATCH CREATE ACTIVITIES COMMAND =====

export interface BatchCreateActivitiesCommand extends ICommand {
  readonly type: 'BatchCreateActivities';
  readonly data: {
    activities: Array<{
      activityType: ActivityType;
      title: string;
      location: string;
      priority: Priority;
      description?: string;
      building?: string;
      zone?: string;
      confidence?: number;
      externalData?: {
        sourceSystem: string;
        originalType: string;
        rawPayload: Record<string, any>;
      };
    }>;
    batchId: string;
    sourceSystem?: string;
  };
}

export interface BatchCreateActivitiesResult {
  batchId: string;
  successCount: number;
  failedCount: number;
  createdActivityIds: string[];
  failures: Array<{
    index: number;
    error: string;
  }>;
  incidentsTriggered: Array<{
    activityId: string;
    incidentId: string;
    rule: string;
  }>;
}

// ===== MERGE ACTIVITIES COMMAND =====

export interface MergeActivitiesCommand extends ICommand {
  readonly type: 'MergeActivities';
  readonly data: {
    primaryActivityId: string;
    activityIdsToMerge: string[];
    mergeReason: string;
    mergedBy: string;
    preserveData?: {
      tags: boolean;
      assignments: boolean;
      timeline: boolean;
    };
  };
}

export interface MergeActivitiesResult {
  primaryActivityId: string;
  mergedActivityIds: string[];
  mergedData: {
    tagsAdded: string[];
    timelineEntriesAdded: number;
    evidenceLinked: string[];
  };
}

// Union type of all activity commands
export type ActivityCommand = 
  | CreateActivityCommand
  | UpdateActivityCommand
  | AssignActivityCommand
  | ArchiveActivityCommand
  | BulkUpdateStatusCommand
  | AddUserTagCommand
  | RemoveUserTagCommand
  | EscalateActivityCommand
  | LinkToIncidentCommand
  | BatchCreateActivitiesCommand
  | MergeActivitiesCommand;