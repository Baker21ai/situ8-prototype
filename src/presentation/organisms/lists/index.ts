export { VirtualizedActivityList } from './VirtualizedActivityList';
export { ActivityListItem, getItemHeight, createItemSizeGetter, ITEM_HEIGHTS } from './ActivityListItem';
export type { ActivityListItemProps } from './ActivityListItem';

// New compound component
export { ActivityList, useActivityListContext } from './ActivityList';
export type {
  ActivityListProps,
  ActivityFilterState,
  ActivityPerformanceMetrics,
  ActivityLayoutMode,
  ActivityViewMode,
  PrioritySegment
} from './ActivityList';