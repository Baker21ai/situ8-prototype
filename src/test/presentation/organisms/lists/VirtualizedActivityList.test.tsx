import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { VirtualizedActivityList } from '../../../../presentation/organisms/lists/VirtualizedActivityList';
import { EnterpriseActivity, ActivityCluster } from '../../../../../lib/types/activity';
import { EnterpriseCardVariant } from '../../../../../components/EnterpriseActivityCard';

// Mock react-window
vi.mock('react-window', () => ({
  FixedSizeList: vi.fn(({ children, itemData, itemCount, itemSize, onItemsRendered, onScroll }) => {
    const renderItem = children;
    return (
      <div data-testid="fixed-size-list">
        {Array.from({ length: Math.min(itemCount, 10) }, (_, index) => (
          <div key={index} data-testid={`list-item-${index}`}>
            {renderItem({ index, style: {}, data: itemData })}
          </div>
        ))}
        <div data-testid="list-controls">
          <button onClick={() => onItemsRendered?.({ visibleStartIndex: 0, visibleStopIndex: 9 })}>
            Trigger onItemsRendered
          </button>
          <button onClick={() => onScroll?.({ scrollOffset: 100 })}>
            Trigger onScroll
          </button>
        </div>
      </div>
    );
  }),
  VariableSizeList: vi.fn(({ children, itemData, itemCount, itemSize, onItemsRendered, onScroll }) => {
    const renderItem = children;
    return (
      <div data-testid="variable-size-list">
        {Array.from({ length: Math.min(itemCount, 10) }, (_, index) => (
          <div key={index} data-testid={`list-item-${index}`}>
            {renderItem({ index, style: {}, data: itemData })}
          </div>
        ))}
        <div data-testid="list-controls">
          <button onClick={() => onItemsRendered?.({ visibleStartIndex: 0, visibleStopIndex: 9 })}>
            Trigger onItemsRendered
          </button>
          <button onClick={() => onScroll?.({ scrollOffset: 100 })}>
            Trigger onScroll
          </button>
        </div>
      </div>
    );
  })
}));

// Mock Badge component
vi.mock('../../../../../components/ui/badge', () => ({
  Badge: ({ children, variant, className }: any) => (
    <div data-testid="badge" className={className}>{children}</div>
  )
}));

// Mock VirtualScrollErrorWrapper
vi.mock('../../../../presentation/atoms/errors', () => ({
  VirtualScrollErrorWrapper: ({ children }: any) => (
    <div data-testid="error-wrapper">{children}</div>
  )
}));

// Mock ActivityListItem
vi.mock('../../../../presentation/organisms/lists/ActivityListItem', () => ({
  ActivityListItem: ({ data, index }: any) => {
    const item = data.items[index];
    return (
      <div 
        data-testid="activity-list-item"
        onClick={() => data.onSelect?.(item)}
      >
        {item.title || 'Untitled'}
      </div>
    );
  },
  getItemHeight: vi.fn(() => 80),
  createItemSizeGetter: vi.fn(() => () => 80),
  ITEM_HEIGHTS: { default: 80, compact: 60, minimal: 40 }
}));

// Mock Lucide icons
vi.mock('lucide-react', () => ({
  AlertTriangle: () => <div data-testid="alert-triangle-icon" />,
  Zap: () => <div data-testid="zap-icon" />,
  Clock: () => <div data-testid="clock-icon" />,
  Shield: () => <div data-testid="shield-icon" />,
  Layers: () => <div data-testid="layers-icon" />
}));

describe('VirtualizedActivityList', () => {
  const mockActivities: EnterpriseActivity[] = [
    {
      id: '1',
      title: 'Critical Security Alert',
      description: 'Unauthorized access detected',
      type: 'security-breach',
      priority: 'critical',
      status: 'detecting',
      timestamp: '2024-01-15T14:30:00Z',
      location: 'Building A',
      zone: 'Zone 1'
    },
    {
      id: '2',
      title: 'High Priority Alert',
      description: 'Suspicious activity',
      type: 'alert',
      priority: 'high',
      status: 'assigned',
      timestamp: '2024-01-15T14:35:00Z',
      location: 'Building B',
      zone: 'Zone 2'
    },
    {
      id: '3',
      title: 'Medium Priority Activity',
      description: 'Regular maintenance',
      type: 'maintenance',
      priority: 'medium',
      status: 'in-progress',
      timestamp: '2024-01-15T14:40:00Z',
      location: 'Building C',
      zone: 'Zone 3'
    }
  ];

  const mockCluster: ActivityCluster = {
    id: 'cluster-1',
    clusterType: 'cluster',
    title: '3 activities in Building A',
    description: 'Multiple events detected',
    location: 'Building A',
    timestamp: '2024-01-15T14:30:00Z',
    lastActivity: '2024-01-15T15:00:00Z',
    count: 3,
    highestPriority: 'critical',
    types: ['security-breach', 'alert'],
    timeRange: '30 minutes',
    activities: mockActivities
  };

  const mockOnSelect = vi.fn();
  const mockOnAction = vi.fn();
  const selectedItems = new Set<string>();

  const defaultProps = {
    items: mockActivities,
    variant: 'detailed' as EnterpriseCardVariant,
    onSelect: mockOnSelect,
    onAction: mockOnAction,
    selectedItems,
    height: 400
  };

  beforeEach(() => {
    vi.clearAllMocks();
    // Mock console.debug to avoid noise in tests
    vi.spyOn(console, 'debug').mockImplementation(() => {});
    vi.spyOn(console, 'log').mockImplementation(() => {});
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe('Rendering', () => {
    it('should render with default props', () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      expect(screen.getByTestId('error-wrapper')).toBeInTheDocument();
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
    });

    it('should render empty state when no items', () => {
      render(<VirtualizedActivityList {...defaultProps} items={[]} />);
      
      expect(screen.getByText('No activities found')).toBeInTheDocument();
      expect(screen.getByText('Try adjusting your filters or search criteria')).toBeInTheDocument();
      expect(screen.getByTestId('alert-triangle-icon')).toBeInTheDocument();
    });

    it('should render activity items', () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      expect(screen.getByText('Critical Security Alert')).toBeInTheDocument();
      expect(screen.getByText('High Priority Alert')).toBeInTheDocument();
      expect(screen.getByText('Medium Priority Activity')).toBeInTheDocument();
    });
  });

  describe('Priority segmentation', () => {
    it('should show priority overview when showPrioritySegments is true', () => {
      render(<VirtualizedActivityList {...defaultProps} showPrioritySegments={true} />);
      
      expect(screen.getByText('1')).toBeInTheDocument(); // Critical count
      expect(screen.getByText('Critical')).toBeInTheDocument();
      expect(screen.getByText('1')).toBeInTheDocument(); // High count  
      expect(screen.getByText('High')).toBeInTheDocument();
      expect(screen.getByText('1')).toBeInTheDocument(); // Medium count
      expect(screen.getByText('Medium')).toBeInTheDocument();
    });

    it('should not show priority overview when showPrioritySegments is false', () => {
      render(<VirtualizedActivityList {...defaultProps} showPrioritySegments={false} />);
      
      expect(screen.queryByText('Critical')).not.toBeInTheDocument();
      expect(screen.queryByText('High')).not.toBeInTheDocument();
    });

    it('should group items by priority', () => {
      render(<VirtualizedActivityList {...defaultProps} showPrioritySegments={true} />);
      
      // Items should be reordered by priority (critical first)
      const listItems = screen.getAllByTestId('activity-list-item');
      expect(listItems[0]).toHaveTextContent('Critical Security Alert');
      expect(listItems[1]).toHaveTextContent('High Priority Alert');
      expect(listItems[2]).toHaveTextContent('Medium Priority Activity');
    });
  });

  describe('Virtualization modes', () => {
    it('should use FixedSizeList for regular activities', () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
      expect(screen.queryByTestId('variable-size-list')).not.toBeInTheDocument();
    });

    it('should use VariableSizeList for clusters', () => {
      render(<VirtualizedActivityList {...defaultProps} items={[mockCluster]} />);
      
      expect(screen.getByTestId('variable-size-list')).toBeInTheDocument();
      expect(screen.queryByTestId('fixed-size-list')).not.toBeInTheDocument();
    });

    it('should use VariableSizeList for evidence variant', () => {
      render(<VirtualizedActivityList {...defaultProps} variant="evidence" />);
      
      expect(screen.getByTestId('variable-size-list')).toBeInTheDocument();
      expect(screen.queryByTestId('fixed-size-list')).not.toBeInTheDocument();
    });

    it('should use VariableSizeList for stream variant', () => {
      render(<VirtualizedActivityList {...defaultProps} variant="stream" />);
      
      expect(screen.getByTestId('variable-size-list')).toBeInTheDocument();
      expect(screen.queryByTestId('fixed-size-list')).not.toBeInTheDocument();
    });
  });

  describe('User interactions', () => {
    it('should call onSelect when item is clicked', async () => {
      const user = userEvent.setup();
      render(<VirtualizedActivityList {...defaultProps} />);
      
      const firstItem = screen.getAllByTestId('activity-list-item')[0];
      await user.click(firstItem);
      
      expect(mockOnSelect).toHaveBeenCalledWith(mockActivities[0]);
    });

    it('should handle keyboard navigation when enabled', () => {
      render(<VirtualizedActivityList {...defaultProps} enableKeyboardNavigation={true} />);
      
      // Test that keyboard event listeners are set up
      // This is mainly testing that the component renders without errors
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
    });

    it('should display keyboard navigation help when enabled', () => {
      render(<VirtualizedActivityList {...defaultProps} enableKeyboardNavigation={true} />);
      
      expect(screen.getByText(/↑↓: Navigate • Enter: Select • Home\/End: Jump/)).toBeInTheDocument();
    });
  });

  describe('Scroll behavior', () => {
    it('should handle scroll events for restoration', async () => {
      render(<VirtualizedActivityList {...defaultProps} enableScrollRestoration={true} />);
      
      const scrollButton = screen.getByText('Trigger onScroll');
      fireEvent.click(scrollButton);
      
      // Should not throw errors
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
    });

    it('should handle items rendered events', async () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      const itemsRenderedButton = screen.getByText('Trigger onItemsRendered');
      fireEvent.click(itemsRenderedButton);
      
      // Should not throw errors and should log debug info
      expect(console.debug).toHaveBeenCalled();
    });
  });

  describe('Performance indicators', () => {
    it('should show performance information', () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      expect(screen.getByText(/Showing 3 activities/)).toBeInTheDocument();
      expect(screen.getByText(/Virtual scrolling enabled/)).toBeInTheDocument();
      expect(screen.getByText(/Memory optimized/)).toBeInTheDocument();
    });

    it('should show item count with proper formatting', () => {
      const manyItems = Array.from({ length: 1500 }, (_, i) => ({
        ...mockActivities[0],
        id: `item-${i}`,
        title: `Activity ${i}`
      }));
      
      render(<VirtualizedActivityList {...defaultProps} items={manyItems} />);
      
      expect(screen.getByText(/Showing 1,500 activities/)).toBeInTheDocument();
    });
  });

  describe('Layout modes', () => {
    it('should handle grid layout mode', () => {
      render(<VirtualizedActivityList {...defaultProps} layoutMode="grid" />);
      
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
    });

    it('should handle horizontal layout mode', () => {
      render(<VirtualizedActivityList {...defaultProps} layoutMode="horizontal" />);
      
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
    });
  });

  describe('Compact mode', () => {
    it('should apply compact mode styling', () => {
      render(<VirtualizedActivityList {...defaultProps} compactMode={true} />);
      
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
    });
  });

  describe('Error handling', () => {
    it('should wrap content in error boundary', () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      expect(screen.getByTestId('error-wrapper')).toBeInTheDocument();
    });

    it('should handle fallback to standard list', () => {
      // This would be triggered by the error boundary in real scenarios
      render(<VirtualizedActivityList {...defaultProps} />);
      
      expect(screen.getByTestId('error-wrapper')).toBeInTheDocument();
    });
  });

  describe('Mixed content handling', () => {
    it('should handle mixed activities and clusters', () => {
      const mixedItems = [mockActivities[0], mockCluster, mockActivities[1]];
      render(<VirtualizedActivityList {...defaultProps} items={mixedItems} />);
      
      expect(screen.getByText('Critical Security Alert')).toBeInTheDocument();
      expect(screen.getByText('3 activities in Building A')).toBeInTheDocument();
      expect(screen.getByText('High Priority Alert')).toBeInTheDocument();
    });
  });

  describe('Custom styling', () => {
    it('should apply custom className', () => {
      const { container } = render(
        <VirtualizedActivityList {...defaultProps} className="custom-list" />
      );
      
      expect(container.firstChild).toHaveClass('custom-list');
    });
  });

  describe('Accessibility', () => {
    it('should be focusable when keyboard navigation is enabled', () => {
      render(<VirtualizedActivityList {...defaultProps} enableKeyboardNavigation={true} />);
      
      const listContainer = screen.getByTestId('fixed-size-list').parentElement;
      expect(listContainer).toHaveAttribute('tabIndex', '0');
    });

    it('should not be focusable when keyboard navigation is disabled', () => {
      render(<VirtualizedActivityList {...defaultProps} enableKeyboardNavigation={false} />);
      
      const listContainer = screen.getByTestId('fixed-size-list').parentElement;
      expect(listContainer).toHaveAttribute('tabIndex', '-1');
    });

    it('should provide proper ARIA labels for priority indicators', () => {
      render(<VirtualizedActivityList {...defaultProps} showPrioritySegments={true} />);
      
      // Priority stats should be properly labeled
      expect(screen.getByText('Critical')).toBeInTheDocument();
      expect(screen.getByText('High')).toBeInTheDocument();
      expect(screen.getByText('Medium')).toBeInTheDocument();
      expect(screen.getByText('Low')).toBeInTheDocument();
    });
  });

  describe('Priority statistics', () => {
    it('should calculate priority statistics correctly', () => {
      render(<VirtualizedActivityList {...defaultProps} showPrioritySegments={true} />);
      
      // Check that priority counts are displayed correctly
      const priorityCards = screen.getAllByText('1');
      expect(priorityCards.length).toBeGreaterThanOrEqual(3); // At least 3 priorities represented
    });

    it('should handle cluster priority statistics', () => {
      render(
        <VirtualizedActivityList 
          {...defaultProps} 
          items={[mockCluster]} 
          showPrioritySegments={true} 
        />
      );
      
      // Should show cluster's highest priority
      expect(screen.getByText('1')).toBeInTheDocument(); // Critical count from cluster
    });
  });

  describe('Item data propagation', () => {
    it('should pass correct item data to list component', () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      // Items should be rendered with their data
      expect(screen.getByText('Critical Security Alert')).toBeInTheDocument();
      expect(screen.getByText('High Priority Alert')).toBeInTheDocument();
    });

    it('should include all necessary props in item data', () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      // The component should render without errors, indicating item data is properly structured
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
    });
  });

  describe('Overscan configuration', () => {
    it('should use appropriate overscan for fixed size list', () => {
      render(<VirtualizedActivityList {...defaultProps} />);
      
      // Component should render (overscan is passed to react-window internally)
      expect(screen.getByTestId('fixed-size-list')).toBeInTheDocument();
    });

    it('should use appropriate overscan for variable size list', () => {
      render(<VirtualizedActivityList {...defaultProps} items={[mockCluster]} />);
      
      // Component should render (overscan is passed to react-window internally)
      expect(screen.getByTestId('variable-size-list')).toBeInTheDocument();
    });
  });

  describe('Snapshot tests', () => {
    it('should match snapshot for fixed size list', () => {
      const { container } = render(<VirtualizedActivityList {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for variable size list with clusters', () => {
      const { container } = render(
        <VirtualizedActivityList {...defaultProps} items={[mockCluster]} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for empty state', () => {
      const { container } = render(
        <VirtualizedActivityList {...defaultProps} items={[]} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
