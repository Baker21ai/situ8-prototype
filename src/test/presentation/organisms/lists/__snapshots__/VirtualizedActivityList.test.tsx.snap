// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VirtualizedActivityList > Snapshot tests > should match snapshot for empty state 1`] = `
<div
  class="flex items-center justify-center h-full text-gray-500 "
>
  <div
    class="text-center"
  >
    <div
      data-testid="alert-triangle-icon"
    />
    <div
      class="text-lg font-medium"
    >
      No activities found
    </div>
    <div
      class="text-sm"
    >
      Try adjusting your filters or search criteria
    </div>
  </div>
</div>
`;

exports[`VirtualizedActivityList > Snapshot tests > should match snapshot for fixed size list 1`] = `
<div
  data-testid="error-wrapper"
>
  <div
    class="flex flex-col h-full "
  >
    <div
      class="flex-shrink-0 p-4"
    >
      <div
        class="grid grid-cols-4 gap-4 mb-4"
      >
        <div
          class="bg-red-50 border border-red-200 rounded-lg p-3 text-center"
        >
          <div
            class="text-2xl font-bold text-red-600"
          >
            1
          </div>
          <div
            class="text-sm text-red-800"
          >
            Critical
          </div>
        </div>
        <div
          class="bg-orange-50 border border-orange-200 rounded-lg p-3 text-center"
        >
          <div
            class="text-2xl font-bold text-orange-600"
          >
            1
          </div>
          <div
            class="text-sm text-orange-800"
          >
            High
          </div>
        </div>
        <div
          class="bg-yellow-50 border border-yellow-200 rounded-lg p-3 text-center"
        >
          <div
            class="text-2xl font-bold text-yellow-600"
          >
            1
          </div>
          <div
            class="text-sm text-yellow-800"
          >
            Medium
          </div>
        </div>
        <div
          class="bg-green-50 border border-green-200 rounded-lg p-3 text-center"
        >
          <div
            class="text-2xl font-bold text-green-600"
          >
            0
          </div>
          <div
            class="text-sm text-green-800"
          >
            Low
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex-1 overflow-hidden"
      tabindex="0"
    >
      <div
        data-testid="fixed-size-list"
      >
        <div
          data-testid="list-item-0"
        >
          <div
            data-testid="activity-list-item"
          >
            Critical Security Alert
          </div>
        </div>
        <div
          data-testid="list-item-1"
        >
          <div
            data-testid="activity-list-item"
          >
            High Priority Alert
          </div>
        </div>
        <div
          data-testid="list-item-2"
        >
          <div
            data-testid="activity-list-item"
          >
            Medium Priority Activity
          </div>
        </div>
        <div
          data-testid="list-controls"
        >
          <button>
            Trigger onItemsRendered
          </button>
          <button>
            Trigger onScroll
          </button>
        </div>
      </div>
    </div>
    <div
      class="flex-shrink-0 p-2 border-t bg-gray-50 text-xs text-gray-600"
    >
      <div
        class="flex items-center justify-between"
      >
        <span>
          Showing 
          3
           activities • Virtual scrolling enabled • Memory optimized
        </span>
        <span>
          ↑↓: Navigate • Enter: Select • Home/End: Jump
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`VirtualizedActivityList > Snapshot tests > should match snapshot for variable size list with clusters 1`] = `
<div
  data-testid="error-wrapper"
>
  <div
    class="flex flex-col h-full "
  >
    <div
      class="flex-shrink-0 p-4"
    >
      <div
        class="grid grid-cols-4 gap-4 mb-4"
      >
        <div
          class="bg-red-50 border border-red-200 rounded-lg p-3 text-center"
        >
          <div
            class="text-2xl font-bold text-red-600"
          >
            1
          </div>
          <div
            class="text-sm text-red-800"
          >
            Critical
          </div>
        </div>
        <div
          class="bg-orange-50 border border-orange-200 rounded-lg p-3 text-center"
        >
          <div
            class="text-2xl font-bold text-orange-600"
          >
            0
          </div>
          <div
            class="text-sm text-orange-800"
          >
            High
          </div>
        </div>
        <div
          class="bg-yellow-50 border border-yellow-200 rounded-lg p-3 text-center"
        >
          <div
            class="text-2xl font-bold text-yellow-600"
          >
            0
          </div>
          <div
            class="text-sm text-yellow-800"
          >
            Medium
          </div>
        </div>
        <div
          class="bg-green-50 border border-green-200 rounded-lg p-3 text-center"
        >
          <div
            class="text-2xl font-bold text-green-600"
          >
            0
          </div>
          <div
            class="text-sm text-green-800"
          >
            Low
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex-1 overflow-hidden"
      tabindex="0"
    >
      <div
        data-testid="variable-size-list"
      >
        <div
          data-testid="list-item-0"
        >
          <div
            data-testid="activity-list-item"
          >
            3 activities in Building A
          </div>
        </div>
        <div
          data-testid="list-controls"
        >
          <button>
            Trigger onItemsRendered
          </button>
          <button>
            Trigger onScroll
          </button>
        </div>
      </div>
    </div>
    <div
      class="flex-shrink-0 p-2 border-t bg-gray-50 text-xs text-gray-600"
    >
      <div
        class="flex items-center justify-between"
      >
        <span>
          Showing 
          1
           activities • Virtual scrolling enabled • Memory optimized
        </span>
        <span>
          ↑↓: Navigate • Enter: Select • Home/End: Jump
        </span>
      </div>
    </div>
  </div>
</div>
`;
